url04f7fcf4e1_peteratt_HDFS_Replication_Simulator_tgz-pJ8-hdfs_replicationsimulator_MainJ8/src/hdfs/replicationsimulator/AllDatanode.java

```diff
@@ -0,0 +1,261 @@
+package hdfs.replicationsimulator;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+
+public class AllDatanode {
+
+	private List<Datanode> datanodes;
+	private boolean isRunning = false;
+	private Queue<Event> toNamenode;
+
+	private Queue<Event> toDatanodes;
+
+	public Daemon hbthread = null; // Heartbeatsender thread
+	public Daemon commandsthread = null; // commands thread
+
+
+	private int heartbeatInterval = Simulator.getHeartbeat();// in ms
+	private int bandwidth		   = Simulator.getBandwidth();//in b/s
+	private int blockSize			= Simulator.getBlockSize()*1000;//in B
+	
+	private long transferTime;
+	//Communication recheck interval
+	private long communicationRecheckInterval = Simulator.getHeartbeat()/2;
+	//Communication hearbeatsender interval
+	private long heartbeatRecheckInterval = Simulator.getHeartbeat()/5;
+
+	// Lock for accessing the namenode command list
+	private Object namenodeLock = new Object();
+
+	// Lock for accessing the datanode command list
+	private Object datanodeLock = new Object();
+
+	boolean checkingPending = false;
+
+	public AllDatanode() {
+
+		datanodes = new ArrayList<Datanode>();
+		toNamenode = Simulator.getToNamenode();
+		toDatanodes = Simulator.getToDatanodes();
+
+		this.hbthread = new Daemon(new HeartbeatSender());
+		this.commandsthread = new Daemon(new CommandsHandler());
+
+		this.bandwidth = Simulator.getBandwidth();
+
+		this.transferTime = (blockSize * 8) / this.bandwidth;
+
+	}
+
+	void addNode(Datanode datanode) {
+		datanodes.add(datanode);
+	}
+
+	Datanode getNode(int id) {
+		return datanodes.get(id);
+	}
+	
+	boolean killNode(int id){
+		return getNode(id).kill();
+	}
+
+	int datanodeListSize() {
+		return datanodes.size();
+	}
+
+	public boolean addToNamenodeQueue(Event e) {
+		synchronized (namenodeLock) {
+			// System.out.println("adding heartbeat to namenode...");
+			boolean result = toNamenode.add(e);
+			// System.out.println("added heartbeat to namenode.");
+			return result;
+		}
+	}
+
+	public boolean addToDatanodeQueue(Event e) {
+		synchronized (datanodeLock) {
+			// System.out.println("adding heartbeat to namenode...");
+			boolean result = toDatanodes.add(e);
+			// System.out.println("added heartbeat to namenode.");
+			return result;
+		}
+	}
+
+	/**
+	 * Starts the datanodes management thread
+	 */
+	public void start() {
+		isRunning = true;
+		hbthread.start();
+		commandsthread.start();
+	}
+
+	/**
+	 * Stops the datanodes management thread
+	 */
+	public void stop() {
+		isRunning = false;
+
+		try {
+			hbthread.join();
+			commandsthread.join();
+
+		} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)